src/compiler/checker.ts

```diff
@@ -36874,7 +36874,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     }
                 }
                 if (contextualSignature && !getReturnTypeFromAnnotation(node) && !signature.resolvedReturnType) {
-                    const returnType = getReturnTypeFromBody(node, checkMode);
+                    const contextualReturnType = getContextualReturnType(node, /*contextFlags*/ undefined);
+                    const returnType = contextualReturnType && contextualReturnType.flags & (TypeFlags.AnyOrUnknown | TypeFlags.Void) ? contextualReturnType : getReturnTypeFromBody(node, checkMode);

```

Comment:
This has a somewhat unwanted side effect - those types become now assigned to `signature.resolvedReturnType`. This shouldn't (usually?) be observable by the user but might be observable by the API users.

There is also a chance that this could break some convoluted scenarios with named functions - when the inner closure has access to a function that I've altered here and calls it directly~. For that reason... perhaps those added lines should skip named functions.